src/data/random/financial.js

```diff
@@ -1,155 +1,127 @@
-            return filter(d.date);
-        });
+    function calculateOHLC(start, price) {
+        var prices = walk
+            .period(granularityInYears())(price);
+        var datum = {
+            date: start,
+            open: prices[0],
+            high: Math.max.apply({}, prices),
+            low: Math.min.apply({}, prices),
+            close: prices[walk.steps()]
+        };
+        datum.volume = volume(datum);
+        return datum;
     }
 
-    var gen = function(days) {
-        var date = startDate,
-            remainingDays = days,
-            result = [],
-            interval;
-
-        do {
-            interval = calculateInterval(date, remainingDays);
-            result = result.concat(dataGenerator(remainingDays, interval.years));
-            date = interval.toDate;
-            remainingDays = days - result.length;
-        }
-        while (result.length < days);
-
-        return result;
+    var financial = function(numPoints) {
+        var stream = makeStream();
+        return stream.take(numPoints);
     };
 
-    gen.mu = function(x) {
-        if (!arguments.length) {
-            return mu;
-        }
-        mu = x;
-        return gen;
-    };
-    gen.sigma = function(x) {
-        if (!arguments.length) {
-            return sigma;
-        }
-        sigma = x;
-        return gen;
-    };
-    gen.startPrice = function(x) {
-        if (!arguments.length) {
-            return startPrice;
+    financial.stream = makeStream;
+
+    function makeStream() {
+        var latest;
+        function getDatum() {
+            if (!latest) {
+                latest = calculateOHLC(new Date(startDate.getTime()), startPrice);
+            } else {
+                latest = calculateOHLC(new Date(latest.date.getTime() + granularity * 1000), latest.close);
+            }
+            return latest;
         }
-        startPrice = x;
-        return gen;
-    };
-    gen.startVolume = function(x) {
-        if (!arguments.length) {
-            return startVolume;
+        function getNextDatum() {
+            var ohlc = getDatum();
+            while (filter && !filter(ohlc)) {
+                ohlc = getDatum();
+            }
+            return ohlc;
         }
-        startVolume = x;
-        return gen;
-    };
-    gen.startDate = function(x) {
+        var stream = {};
+        stream.next = function() {
+            return getNextDatum();
+        };
+        stream.take = function(numPoints) {
+            var point;
+            var data = [];
+            for (point = 0; point < numPoints; point += 1) {
+                data.push(getNextDatum());
+            }
+            return data;
+        };
+        stream.until = function(comparison) {
+            var data = [];
+            if (!comparison || latest && comparison(latest)) {
+                return data;
+            }
+            var ohlc = getNextDatum();
+            while (comparison) {
```

Comment:
Yeah I don't much like how this turned out - I think the takeUntil behavior would work, would just have to move setting the latest point out of `getDatum`, otherwise the next call to a stream method after an `until` call returns the date unexpectedly incremented.